source/common/formatter/substitution_formatter.cc

```diff
@@ -148,11 +148,23 @@ std::string JsonFormatterImpl::format(const Http::RequestHeaderMap& request_head
   return absl::StrCat(log_line, "\n");
 }
 
+StructFormatter::StructFormatter(const ProtobufWkt::Struct& format_mapping, bool preserve_types,
+                                 bool omit_empty_values,
+                                 const std::vector<CommandParserPtr>& commands)
+    : omit_empty_values_(omit_empty_values), preserve_types_(preserve_types),
+      empty_value_(omit_empty_values_ ? EMPTY_STRING : DefaultUnspecifiedValueString),
+      commands_(commands), struct_output_format_(toFormatMapValue(format_mapping)) {
+  // Note: to avoid copying commands we save the ref in an optional for toFormatMapValue and related
+  // methods to use. Once those are done we need to unset the optional to avoid accidentally
+  // accessing a dangling reference.
+  commands_ = absl::nullopt;
```

Comment:
I'd add a note about why we don't just pass it through the call chain as well, as that'd be my first question reading this

Maybe even a TODO about using a builder pattern to construct the format output, e.g. `FormatBuilder(commands).buildFormatMapValue(format_mapping)`

Reply:
Or if you want to just use a builder/factory pattern directly that would also work :) I realized it would be a nice way to ensure that the ref doesn't outlive the ctor and not have to pass it through all the helper functions

Reply:
Hmm if you feel strongly about it, I might as well add a FormatBuilder class and use that instead. No need to accumulate another TODO :-)

Reply:
Done.